docs/adr/002-monolith-architecture-with-modular-subsystems.md

```diff
@@ -0,0 +1,75 @@
+# Monolith Architecture with Modular Subsystems
+
+## Context and Problem Statement
+
+The Nuts Knooppunt needs to integrate multiple subsystems including authentication (Nuts node), authorization (OPA), addressing (mCSD), localization, and consent management. Currently, these would require deploying and managing multiple containers, which adds operational complexity for parties deploying a Knooppunt. We need to decide on the overall architecture approach to favor simplicity and "plug and play" deployment while maintaining flexibility for future scaling.
+
+## Considered Options
+
+* **Multiple containers/microservices** - Each subsystem as a separate container requiring individual deployment and management
+* **Pure monolith with function calls** - Single container with tightly integrated subsystems communicating via direct function calls for maximum efficiency
+* **Jaeger-style all-in-one** - Single binary that can run either in all-in-one mode (all components) or with subcommands to run individual components separately (e.g., `jaeger-all-in-one` vs `jaeger-collector`, `jaeger-query`). This provides deployment flexibility while maintaining a single codebase
+
+## Decision Outcome
+
+Chosen option: "Hybrid monolith combining all-in-one deployment with HTTP communication", which combines elements from multiple approaches:
+
+**From the Jaeger-style all-in-one approach:**
+* Single binary deployment
+* Component lifecycle management allowing selective enablement
+* Potential for future subcommands to run components individually
+
+**From the monolith with HTTP communication approach:**
+* HTTP-based internal communication instead of function calls
+* Clear API boundaries between subsystems
+* Each subsystem mounts its handlers on shared HTTP servers
+
+**Key benefits of this combined approach:**
+* **Simplified deployment** - Default all-in-one mode for easy "plug and play" deployment
+* **Deployment flexibility** - Components can be selectively enabled/disabled via configuration
+* **Clear boundaries** - HTTP APIs prevent tight coupling and enable testing/mocking
+* **Future-proof** - Can evolve to run components separately or extract to microservices
+* **Technical feasibility** - All components available in Go with shared lifecycle interface
+
+This hybrid approach gives us the simplicity of a monolith with the architectural flexibility of a more modular system.
+
+## Implementation Details
+
+### HTTP Interface Design
+
+**Two separate HTTP servers:**
+- **Public API** (port 8080) - External-facing endpoints
+- **Private API** (port 8081) - Internal/health endpoints
+
+**URL structure example:**
+- `host:8080/{component}/*` - Public component endpoints (e.g., `/auth/*`, `/addressing/*`)
+- `host:8081/health` - Health check endpoint
+- `host:8081/admin/*` - Administrative interfaces
+
+### Inter-subsystem Communication
+
+**Deliberate choice: HTTP over in-process calls**
```

Comment:
We can always optimize this in the future by making calls to embedded subsystems local in-process calls, if the subsystems allows it. I can image that would work for OPA and Nuts node. But as it adds complexity, let not take that in scope for now (since then we'd have to test both variants, embedded and external).